observability/test/run_o11y_tests.py

```diff
@@ -0,0 +1,1174 @@
+
+def parse_args() -> argparse.Namespace:
+    argp = argparse.ArgumentParser(description='Run Observability integration tests')
+    argp.add_argument('--server_lang', required=True, type=lambda s:SupportedLang(s),
+                      help='Server language')
+    argp.add_argument('--client_lang', required=True, type=lambda s:SupportedLang(s),
+                      help='Client language')
+    argp.add_argument('--job_mode', required=True, type=lambda s:JobMode(s),
+                      help='Job mode')
+    argp.add_argument('--test_case', required=True, type=lambda s:ObservabilityTestCase(s),
+                      help='Test case to run')
+    argp.add_argument('--port', required=False, default='',
+                      help='Port number TestService server should run on')
+    argp.add_argument('--gke_resource_identifier', required=False, default='',
+                      help='GKE resource identifier, most likely a suffix')
+    args = argp.parse_args()
+    logger.debug('Parsed args: ' + str({k: str(v) for k, v in vars(args).items()}))
+    return args
+
+class TestRunner:
+    args: argparse.Namespace
+    job_name: str
+    exit_status: int
+
+    def __init__(self, args: argparse.Namespace) -> None:
+        self.args = args
+        self.job_name = '%s:%s:%s' % (self.args.server_lang,
+                                      self.args.client_lang,
+                                      self.args.test_case)
+        self.exit_status = 0
+
+    def set_exit_status(self, status: int) -> None:
+        self.exit_status = status
+
+    def write_sponge_xml(self) -> None:
+        sponge_dir = TestUtil.get_sponge_log_dir(self.args.job_mode, self.job_name)
+        with open(os.path.join(sponge_dir, 'sponge_log.xml'), 'w') as f:
+            f.write(str(self))
+
+    def __str__(self) -> str:
+        timestamp = datetime.now(timezone.utc).isoformat()
+        res = '<testsuites>\n'
+        res += '  <testsuite errors="0" failures="%d" name="grpc-o11y-%s" ' % (
+            self.exit_status, self.job_name)
+        res += 'package="grpc-o11y-tests" timestamp="%s">\n' % timestamp
+        res += '  </testsuite>\n</testsuites>'
+        return res
+
+    def run(self) -> None:
+        try:
+            impl = TestCaseImpl(self)
+            test_case_func = getattr(impl, str(self.args.test_case))
+            logger.info('Executing TestCaseImpl.%s()' % str(self.args.test_case))
+            test_case_func()
+        except Exception:
+            self.set_exit_status(1)
+            logger.error(traceback.format_exc())
+        finally:
+            self.write_sponge_xml()
+            logger.info("Test case '%s' %s." % (self.args.test_case,
+                                             'failed' if self.exit_status else 'passed'))
+            sys.exit(self.exit_status)
+
+class GKETestEnv:
+    args: argparse.Namespace
+    k8s_core_v1: kubernetes_client.CoreV1Api
+    k8s_client: kubernetes_client.ApiClient
```

Comment:
We just learned an interesting quirk about this library that you'll want to incorporate into your code: https://github.com/grpc/grpc/pull/32210

You should only see an issue with this however if:

1. The tests exceed ~45 minutes
2. or the GCP token lifetime is decreased on the backend

Neither of these is likely to happen immediately, but it will be real headache if/when it does.

Reply:
I had made a note of this. Thanks for the pointer.